dns/dns_cache.go

```diff
@@ -0,0 +1,61 @@
+package dns
+
+import (
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/xvzc/SpoofDPI/util"
+)
+
+type DNSCacheEntry struct {
+	ip           string
+	expiry_timer *time.Timer
+}
+
+type DNSCache struct {
+	cacheMap  map[string]DNSCacheEntry
+	cacheLock sync.RWMutex
+}
+
+var dnsCache *DNSCache
+var once sync.Once
+
+func GetCache() *DNSCache {
+	once.Do(func() {
+		dnsCache = &DNSCache{
+			cacheMap:  make(map[string]DNSCacheEntry),
+			cacheLock: sync.RWMutex{},
+		}
+	})
+	return dnsCache
+}
+
+// interface function for the inner map basically.
+func (d *DNSCache) Set(key string, value string) {
+	d.cacheLock.Lock()
+	defer d.cacheLock.Unlock()
+	if _, ok := d.cacheMap[key]; ok {
+		return
+	}
+
+	d.cacheMap[key] = DNSCacheEntry{
+		ip: value,
+		expiry_timer: time.AfterFunc(time.Duration(*util.GetConfig().DNSCacheExpiry)*time.Second, func() {
+			d.cacheLock.Lock()
+			defer d.cacheLock.Unlock()
+			delete(d.cacheMap, key)
+		}),
+	}
+}
+
+func (d *DNSCache) Get(key string) (string, error) {
```

Comment:
What do you think about returning `(string, bool)`, not `(string, error)`. There is no way this function can return any other error than `cache missed`, so you can just return `false` on miss and `true` on hit. This way `Get` function will be just like get in maps:
```go
map_ := map[string]string{"google.com", "dnsValue"}
facebookDns, ok := map_["facebook.com"]
if ok{
    // hit
}
// miss
```